Ginger/GingerCoreCommon/Repository/BusinessFlowLib/ActivitiesGroup.cs

```diff
@@ -204,49 +204,143 @@ public override string ItemName
+
+                    List<Activity> existingActivities = new List<Activity>();
+
+                    int exActCount = activitiesGroupInstance.ActivitiesIdentifiers.Count;
+                    for (int i = exActCount-1; i >= 0; i--)
                     {
-                        //keep other parts
-                        newInstance.ActivitiesIdentifiers = activitiesGroupInstance.ActivitiesIdentifiers;
+                        ActivityIdentifiers actIDexist = activitiesGroupInstance.ActivitiesIdentifiers[i];
+
+                        Activity exAct = currentBF.Activities.Where(g => g.Guid == actIDexist.ActivityGuid).FirstOrDefault();
+
+                        if (exAct == null)
+                        {
+                            exAct = currentBF.Activities.Where(g => g.ParentGuid == actIDexist.ActivityGuid).FirstOrDefault();
+                        }
+
+                        if (exAct != null)
+                        {
+                            // Add to the list of deleted activities
+                            existingActivities.Add(exAct);
+
+                            // Remove the activity from the Business Flow
+                            currentBF.DeleteActivity(exAct);
+                        }
                     }
-                    if (hostItem != null)
+
+                    // Add the activities to the Business Flow in sequence they appear in the Updated Shared Group
+                    foreach (ActivityIdentifiers actID in newInstance.ActivitiesIdentifiers)
                     {
-                        //replace old instance object with new
-                        int originalIndex = ((BusinessFlow)hostItem).ActivitiesGroups.IndexOf(activitiesGroupInstance);
-                        ((BusinessFlow)hostItem).ActivitiesGroups.Remove(activitiesGroupInstance);
-                        ((BusinessFlow)hostItem).SetUniqueActivitiesGroupName(newInstance);
-                        ((BusinessFlow)hostItem).ActivitiesGroups.Insert(originalIndex, newInstance);
-                        ((BusinessFlow)hostItem).AttachActivitiesGroupsAndActivities();
+                        Activity updatedAct = null;
+
+                        // Activity still exist in the group, thus re-add the same activity to the group
+                        updatedAct = existingActivities.Where(a => a.Guid == actID.ActivityGuid).FirstOrDefault();
+
+                        // In case, group was Replaced/Overwritten
+                        if (updatedAct == null)
+                        {
+                            updatedAct = existingActivities.Where(a => a.ParentGuid == actID.ActivityGuid).FirstOrDefault();
+                        }
+
+                        if (updatedAct != null)
+                        {
+                            currentBF.AddActivity(updatedAct, activitiesGroupInstance, insertIndex);         //  .Activities.Insert(firstActivityIndex, updatedAct);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this commented out code.](https://app.codacy.com/app/Ginger-Automation/Ginger/pullRequest?prid=4582993)